cli/cmd/add/component-version/cmd.go

```diff
@@ -0,0 +1,359 @@
+	return prov.pluginManager.InputRegistry.GetSourceInputPlugin(ctx, src.Input)
+}
+
+func (prov *constructorProvider) GetResourceRepository(ctx context.Context, resource *constructorruntime.Resource) (constructor.ResourceRepository, error) {
+	return prov.pluginManager.ResourcePluginRegistry.GetResourcePlugin(ctx, resource.Access)
+}
+
+func (prov *constructorProvider) GetTargetRepository(ctx context.Context, _ *constructorruntime.Component) (constructor.TargetRepository, error) {
+	plugin, err := prov.pluginManager.ComponentVersionRepositoryRegistry.GetPlugin(ctx, prov.targetRepoSpec)
+	if err != nil {
+		return nil, fmt.Errorf("getting plugin for repository %q failed: %w", prov.targetRepoSpec, err)
+	}
+	var creds map[string]string
+	identity, err := plugin.GetComponentVersionRepositoryCredentialConsumerIdentity(ctx, prov.targetRepoSpec)
+	if err == nil {
+		if creds, err = prov.graph.Resolve(ctx, identity); err != nil {
+			return nil, fmt.Errorf("getting credentials for repository %q failed: %w", prov.targetRepoSpec, err)
+		}
+	}
+	repo, err := plugin.GetComponentVersionRepository(ctx, prov.targetRepoSpec, creds)
+	if err != nil {
+		return nil, fmt.Errorf("getting component version repository for %q failed: %w", prov.targetRepoSpec, err)
+	}
+	return repo, err
+}
+
+func registerConstructorProgressTracker(cmd *cobra.Command, options constructor.Options) (opts constructor.Options, stop func(), err error) {
+	format, err := enum.Get(cmd.Flags(), log.FormatFlagName)
+	if err != nil {
+		return opts, nil, fmt.Errorf("failed to get the log format from the command flag: %w", err)
+	}
+
+	switch format {
+	case log.FormatText:
+		pw := progress.NewWriter()
+		pw.SetOutputWriter(cmd.OutOrStdout())
+		pw.SetUpdateFrequency(100 * time.Millisecond)
+		pw.SetAutoStop(false)
+		trackers := map[string]*progress.Tracker{}
+		options.OnStartComponentConstruct = func(_ context.Context, component *constructorruntime.Component) error {
+			key := component.Name + "/" + component.Version
+			tracker := &progress.Tracker{
+				Message: "component " + key,
+				Total:   1,
+				Units: progress.Units{
+					Formatter: func(value int64) string {
+						base := fmt.Sprintf("%d component version", value)
+						if value > 1 {
+							base += "s"
+						}
+						return base
+					},
+				},
+			}
+			trackers[key] = tracker
+			pw.AppendTracker(tracker)
+			return nil
+		}
+		options.OnEndComponentConstruct = func(_ context.Context, descriptor *descriptor.Descriptor, err error) error {
+			if err != nil {
+				return nil
+			}
+			key := descriptor.Component.Name + "/" + descriptor.Component.Version
+			tracker := trackers[key]
+			tracker.UpdateMessage(tracker.Message + " constructed")
+			tracker.Increment(1)
+			tracker.MarkAsDone()
+			return nil
+		}
+		// TODO Add Resource and Source tracking in more detail
+		go func() {
+			for _, tracker := range trackers {
+				tracker.Start()
```

Comment:
Does this need some kind of stop mechanism / timeout?

Reply:
the starting itself does not block, so imo it doesnt need a timeout. Will double check